### PR TITLE
Apply 'ColorizeVolume' to every frame of a sequence

### PR DESCRIPTION
**Pull request by henrykrumb**
Addresses issue https://github.com/PerkLab/SlicerSandbox/issues/26 .

If the selected input volume and segmentation belong to a sequence, a colorized volume is created for each frame of that sequence. The user is prompted beforehand if they really want every item of the sequence to be processed.

https://github.com/PerkLab/SlicerSandbox/assets/25531076/f2cea29e-0d85-490f-a211-76c2cebdf8ba

### CODE REVIEW

**Comment by lassoan** [comment]
Thanks a lot, this will be a very useful feature! I had trouble reviewing the actual changes because they are masked by all the formatting changes. Would you be able to move formatting changes into a separate commit? (e.g., use the code formatter on the original source code, commit; then add all the changes in a second commit => then in te second commit we should only have the functional changes). Thank you!

File: ColorizeVolume/ColorizeVolume.py [hunk 1]
```diff
@@ -97,13 +109,15 @@ def setup(self) -> None:
         """
         ScriptedLoadableModuleWidget.setup(self)
 
-        if not hasattr(vtkAddon, 'vtkImageLabelDilate3D'):
-            slicer.util.errorDisplay("This module requires a more recent version of Slicer. Please download and install latest Slicer Preview Release.")
+        if not hasattr(vtkAddon, "vtkImageLabelDilate3D"):
+            slicer.util.errorDisplay(
+                "This module requires a more recent version of Slicer. Please download and install latest Slicer Preview Release."
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:114]: There is no need for breaking this single line command into 3 lines. It is maybe 10-15 characters narrower, but 3x longer.

File: ColorizeVolume/ColorizeVolume.py [hunk 2]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
         # Connections
 
         # These connections ensure that we update parameter node when scene is closed
-        self.addObserver(slicer.mrmlScene, slicer.mrmlScene.StartCloseEvent, self.onSceneStartClose)
-        self.addObserver(slicer.mrmlScene, slicer.mrmlScene.EndCloseEvent, self.onSceneEndClose)
+        self.addObserver(
+            slicer.mrmlScene, slicer.mrmlScene.StartCloseEvent, self.onSceneStartClose
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:137]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 3]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
+        )
+        self.addObserver(
+            slicer.mrmlScene, slicer.mrmlScene.EndCloseEvent, self.onSceneEndClose
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:140]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 4]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
+        )
 
-        self.ui.outputRgbaVolumeSelector.connect('currentNodeChanged(vtkMRMLNode*)', self.onOutputRgbaVolumeSelected)
+        self.ui.outputRgbaVolumeSelector.connect(
+            "currentNodeChanged(vtkMRMLNode*)", self.onOutputRgbaVolumeSelected
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:144]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 5]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
-        self.ui.volumeRenderingLevelWidget.connect('valueChanged(double)', self.onUpdateVolumeRenderingTransferFunction)
-        self.ui.volumeRenderingWindowWidget.connect('valueChanged(double)', self.onUpdateVolumeRenderingTransferFunction)
-        self.ui.volumeRenderingOpacityWidget.connect('valueChanged(double)', self.onUpdateVolumeRenderingTransferFunction)
-        self.ui.volumeRenderingGradientOpacityCheckBox.connect('toggled(bool)', self.onUpdateVolumeRenderingTransferFunction)
-        self.ui.volumeRenderingGradientOpacityLevelWidget.connect('valueChanged(double)', self.onUpdateVolumeRenderingTransferFunction)
-        self.ui.volumeRenderingGradientOpacityWindowWidget.connect('valueChanged(double)', self.onUpdateVolumeRenderingTransferFunction)
+        self.ui.applyButton.connect("clicked(bool)", self.onApplyButton)
+        self.ui.showVolumeRenderingButton.connect("clicked(bool)", self.onShowButton)
+        self.ui.resetVolumeRenderingSettingsButton.connect(
+            "clicked(bool)", self.onResetVolumeRenderingSettingsButton
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:151]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 6]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
+        )
+        self.ui.volumeRenderingSettingsButton.connect(
+            "clicked(bool)", self.onVolumeRenderingSettingsButton
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:154]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 7]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
+        )
+        self.ui.resetToDefaultsButton.connect(
+            "clicked(bool)", self.onResetToDefaultsButton
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:157]: no need for breaking into 3 lines

File: ColorizeVolume/ColorizeVolume.py [hunk 8]
```diff
@@ -119,24 +133,48 @@ def setup(self) -> None:
...
+        )
+
+        self.ui.volumeRenderingLevelWidget.connect(
+            "valueChanged(double)", self.onUpdateVolumeRenderingTransferFunction
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:161]: OK, I stop commenting on this, just revert all these breakups of single-line commands into 3 lines. It just does not make any sense.

File: ColorizeVolume/ColorizeVolume.py [hunk 9]
```diff
@@ -522,22 +719,19 @@ def updateVolumeRenderingOpacityTransferFunctions(self):
...
-                (100.0, 1.0)
-                ]
+            nodes = [(-10.0, 1.0), (0.0, 1.0), (90.0, 1.0), (100.0, 1.0)]
         gradientOpacityTransferFunction = vrProp.GetGradientOpacity()
         if gradientOpacityTransferFunction.GetSize() == len(nodes):
             for index, node in enumerate(nodes):
-                gradientOpacityTransferFunction.SetNodeValue(index, [node[0], node[1], 0.5, 0.0])  # extra values: midpoint, sharpness
+                gradientOpacityTransferFunction.SetNodeValue(
+                    index, [node[0], node[1], 0.5, 0.0]
+                )  # extra values: midpoint, sharpness
```
Comment (lassoan) [inline ColorizeVolume/ColorizeVolume.py:734]: comment misplaced (it refers to values 0.5, 0.0)

**Review by lassoan — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution. I really appreciate that you developed and shared this new feature - it looks great!

It is also nice that you are improving the style of the code (using " consistently, fixing usage of spaces, etc.). However, the amount of style changes are so excessive and the breaking up of quite short lines into several lines was so annoying and made the code harder to read, that I just could not force myself to complete the meaningful part of the review.

Could you please move all style changes into a separate commit and change your autoformatter settings to not to break up lines that are shorter than 160 characters?

If you feel that you have already done enough by contributing this work and would rather not spend any time on this then let me know, I completely understand. Then I'll try to find the time and patience to deal with this as is.

**Comment by henrykrumb** [comment]
Hi Andras,

I am really sorry for the inconvenience. It is a habit of mine to use black as an auto-formatter on every Python file I edit, and I didn't notice before committing.

I will revert the changes made by the auto-formatter (many of which don't even make much sense, as you pointed out) and create another commit with the changes I consider final.

**Comment by henrykrumb** [comment]
I just reverted the changes made by the auto-formatter and kept the imports local to functions (thank you for the hint, I will consider it in the future!).

I also tested this commit with a version of vanilla Slicer with my modified version of TotalSegmentator (https://github.com/lassoan/SlicerTotalSegmentator/pull/56) and the example CTA sequence and it still works, so I guess it didn't break anything.

Again, I'm sorry for the inconvenience and I hope you didn't take too much time to review the formatting.

**Review by lassoan — APPROVED** [review]
I've made a few simplifications (use Sequences module logic to get sequence browser, moved confirmation popup to module GUI, etc.). It all looks good now. Thanks again for your contribution!

**Pull request merged**